### ISSUE TITLE
nvvm built-in funs translates to spirv built-in funs

### ISSUE THREAD

**Issue by easonwang27** [vortexgpgpu/NVPTX-SPIRV-Translator#1]
hello,
Is the translation of nvvm built-in funs into spirv standard built-in funs implemented here?

Does it apply to all translations of nvvm built-in funs?

ths！

**Comment by drcut** [comment]
Sorry for the late reply.
Yes, for example, we support thread index in https://github.com/vortexgpgpu/NVPTX-SPIRV-Translator/blob/master/lib/SPIRV/OCL20ToSPIRV.cpp#L1332. 
The whole process is: we utilize the existing OpenCL->SPIR-V translation to implement NVVM->SPIR-V. So in most cases, for each CUDA feature, you only need to invoke the corresponding translation function for the corresponding features in OpenCL.

**Comment by easonwang27** [comment]
Thank you very much for your reply. I am currently engaged in pocl-related work, and I am very interested in what you are doing now.

I have also seen your paper and I am excited about your ideas. At present, I have simulated your entire process on the X86 platform (pocl device basic). I am really excited.

Do you have a plan for porting other built-in functions of cuda?

**Comment by drcut** [comment]
> Thank you very much for your reply. I am currently engaged in pocl-related work, and I am very interested in what you are doing now.
> 
> I have also seen your paper and I am excited about your ideas. At present, I have simulated your entire process on the X86 platform (pocl device basic). I am really excited.
> 
> Do you have a plan for porting other built-in functions of cuda?

Yes, however, for some built-in functions, to be executed on x86, we have to not only parse and translate the NVPTX built-in functions, but also support the corresponding parts on x86. For example, if you want to support tensorcore, you have to parse the tensorcore part in NVPTX IR, and also support a gemm on x86. 
We are now working on a runtime system, which includes a runtime library for these built-in function. Once it reaches a milestone, we will open source it and update the translator according.
Once again, thanks for your interest. If you have any other questions, please feel free to contact me.

**Comment by easonwang27** [comment]
I personally understand that the conversion tool NVPTX-SPIRV-Translator is to better port cuda applications to the opencl framework.
NVPTX-SPIRV-Translator can convert the kernel part very well, so do you have any better method for the host part code of cuda? Convert cuda runtime API to opencl runtime API.
Or do not need to be converted, there are other ways to complete it?

Thanks a lot!

**Comment by easonwang27** [comment]
I personally understand that the conversion tool NVPTX-SPIRV-Translator is to better port cuda applications to the opencl framework.
NVPTX-SPIRV-Translator can convert the kernel part very well, so do you have any better method for the host part code of cuda? Convert cuda runtime API to opencl runtime API.
Or do not need to be converted, there are other ways to complete it?

Thanks a lot!

> > Thank you very much for your reply. I am currently engaged in pocl-related work, and I am very interested in what you are doing now.
> > I have also seen your paper and I am excited about your ideas. At present, I have simulated your entire process on the X86 platform (pocl device basic). I am really excited.
> > Do you have a plan for porting other built-in functions of cuda?
> 
> Yes, however, for some built-in functions, to be executed on x86, we have to not only parse and translate the NVPTX built-in functions, but also support the corresponding parts on x86. For example, if you want to support tensorcore, you have to parse the tensorcore part in NVPTX IR, and also support a gemm on x86. We are now working on a runtime system, which includes a runtime library for these built-in function. Once it reaches a milestone, we will open source it and update the translator according. Once again, thanks for your interest. If you have any other questions, please feel free to contact me.


I personally understand that the conversion tool NVPTX-SPIRV-Translator is to better port cuda applications to the opencl framework.
NVPTX-SPIRV-Translator can convert the kernel part very well, so do you have any better method for the host part code of cuda? Convert cuda runtime API to opencl runtime API.
Or do not need to be converted, there are other ways to complete it?

Thanks a lot!

**Comment by easonwang27** [comment]
I personally understand that the conversion tool NVPTX-SPIRV-Translator is to better port cuda applications to the opencl framework.
NVPTX-SPIRV-Translator can convert the kernel part very well, so do you have any better method for the host part code of cuda? Convert cuda runtime API to opencl runtime API.
Or do not need to be converted, there are other ways to complete it?

Thanks a lot!

**Comment by drcut** [comment]
> I personally understand that the conversion tool NVPTX-SPIRV-Translator is to better port cuda applications to the opencl framework. NVPTX-SPIRV-Translator can convert the kernel part very well, so do you have any better method for the host part code of cuda? Convert cuda runtime API to opencl runtime API. Or do not need to be converted, there are other ways to complete it?
> 
> Thanks a lot!

There is a project (https://ui.adsabs.harvard.edu/abs/2011CoPhC.182.1093H/abstract) focused on translating CUDA to OpenCL, but it's quite old. In this work, the tools use regular expression to do a source-to-source translation. However, in the recent years, CUDA's API becomes much more complex, especially for the host part (cudaGraph, cooperative group, unified memory).  I believe CUDA's updating is much more faster than OpenCL and SPIR-V. Thus, I don't think it will be a good idea to follow the CUDA->SPIR-V->OpenCL path, as some features maynot existing in SPIR-V or OpenCL. 
We are currently working on executing CUDA to a given backend devices, thus, we just directly parse and translate the CUDA to NVVM IR, and uses the translated NVVM IR to generate executable files with runtime library we manually implemented for that devices. 
In summary, I believe we need an inclusive IR to support all CUDA features, OpenCL and SPIR-V cannot satisfy it currently.

**Comment by easonwang27** [comment]
Thank you for your guidance！
